lesson-1/services/habit.service.js

```diff
@@ -0,0 +1,159 @@
+import { getAll, create, deleteById, updateById, addDoneMark } from '../models/habit.model.js';
+import { ErrorCodes, AppError } from '../utils/errors.js';
+import { validateEnum, validateString } from '../utils/validate.js';
+import { calculateHabitProgress } from '../utils/stat.js';
+
+const validateFrequency = (freq) => validateEnum(freq, ['daily', 'weekly', 'monthly'], { message: 'Invalid --freq parameter. Use "daily", "weekly", or "monthly".' });
+const validateName = (name) => validateString(name, { isRequired: true, message: 'Invalid --name parameter. It must be a non-empty string.' });
+const validateId = (id) => validateString(id, { isRequired: true, message: 'Invalid --id parameter. It must be a non-empty string.' });
+
+const removeOptionalEmptyParams = (params, optionalFields) => {
+  return Object.entries(params).reduce((acc, [name, value]) => {
+    if (optionalFields.includes(name) && value == undefined) {
+      return acc;
+    }
+
+    return {
+      ...acc,
+      [name]: value,
+    };
+  }, {});
+};
+
+const paramsValidators = {
+  name: validateName,
+  freq: validateFrequency,
+  id: validateId,
+};
+
+const validateParameters = (params) => {
+  for (const [param, value] of Object.entries(params)) {
+    const validator = paramsValidators[param];
+
+    if (validator) {
+      validator(value);
+    }
+  }
+};
+
+const hideHabitFields = (habits, fields) => habits.map((habit) => {
+  return Object.entries(habit).reduce((acc, [field, value]) => {
+    if (fields.includes(field)) {
+      return acc;
+    }
+
+    return {
+      ...acc,
+      [field]: value,
+    };
+  }, {});
+});
+
+async function loadAllHabits() {
+  try {
+    const habits = await getAll();
+
+    return hideHabitFields(habits, ['doneMarks']);
+  } catch (error) {
+    throw new AppError({
+      error,
+      code: ErrorCodes.READ_HABITS_ERROR,
+    });
+  }
+}
```

Comment:
a little cleaner way, IMHO
```suggestion
const Try = async (fn, code) => {
  try {
    return await fn();
  } catch (error) {
    throw new AppError({
      error,
      code,
    });
  }
}

const loadAllHabits = () => Try(async () => {
  const habits = await getAll();

  return hideHabitFields(habits, ['doneMarks']);
  }, ErrorCodes.READ_HABITS_ERROR)
```